### DISCUSSION TITLE
Server Tag Automation

### DISCUSSION THREAD

**Discussion by greggarner-io** [degen-workgroup/DEGEN#6] [Ideas]
I took a look through some of the chat and it seems like server tag automation is a great place to start making progress.

Which roles do we want to automate? Should we create a list of the roles (with IDs)?
- Guilds?
- Geographies?

How do we want to trigger these role assignments? 
- Sending messages to the bot?
- Reactions to a message?
- Typing a command in a specific channel? (or all)

If using a claiming strategy, do we need to verify the user is allowed to claim that role?

*Initial response from @nonsensecodes before moving to GitHub*
"The verification aspect is easily handled in the bot framework, as we can specify who can access what tags. As for triggers, I think we should have both one message that people can react to, and a command that people can call from anywhere. “Spammy stuff” seems the popular place for bot commands, but just like tipping, it can happen from wherever."

**Comment by lvnacy** [comment]
Yes, we’ll def need a list of all available tags. The Discord.js library allows for granular access to the Discord API, so we don’t need the hardcoded IDs, which is nice, but the IDs will be useful nevertheless. For the initial feature, I think we should have two levels: the “everyone” level which would include everyone in level 1 and above to claim and hand out guild and locale tags, and the admin level for admins to assign higher level tags.

Role verification is super easy in the Commando framework. Prior to the actual run function for the command, there’s a place for a `hasPermission` function for role validation. We can also specify admins and guild owners in a separate js file for ease of reference. Commando is pretty powerful.

I do not want to focus on this at the moment, but something to consider once this is launched: how to update the bot with new tags. I don’t think this would be difficult to implement, but let’s get the basic functionality working first.

**Comment by SlinkyPotato** [comment]
One thing that might help with onboarding new members is an interactive guest pass feature. Command can be something like

$guest-pass @username

- mentioned user is given role "Guest Pass"
- bot sends a message greeted user and notifying of new role
- expiration timestamp of guest privileges is stored in db
- 15 min reminder of guest pass expiration is send to user
- guest pass is removed at time of expiration

This is the code I have so far https://github.com/BanklessDAO/discord-bots/tree/feature/add-server-role 

Some thoughts:
- should expiration of guest pass be the same for all users? (i.e hardcoded value)
- I imagine the #get-involved channel being used mostly with this command so  bot messages here should be minimal
- there is a 50 api call per second rate limit on discord, so a lot of the automation should have wait timers or we might see issues similar to POAP?

@nonsensecodes @greggarner-io
